web3.js/src/transaction.js

```diff
@@ -350,70 +381,39 @@ export class Transaction {
    * The Transaction must be assigned a valid `recentBlockhash` before invoking this method
    */
   sign(...signers: Array<Account>) {
-    this.signPartial(...signers);
+    if (signers.length === 0) {
+      throw new Error('No signers');
+    }
+
+    this.signatures = signers.map(signer => ({
+      signature: null,
+      publicKey: signer.publicKey,
+    }));
+
+    this.partialSign(...signers);
   }
 
   /**
-   * Partially sign a Transaction with the specified accounts.  The `Account`
-   * inputs will be used to sign the Transaction immediately, while any
-   * `PublicKey` inputs will be referenced in the signed Transaction but need to
-   * be filled in later by calling `addSigner()` with the matching `Account`.
+   * Partially sign a transaction with the specified accounts. All accounts must
+   * correspond to a public key that was previously provided to `setSigners`.
    *
-   * All the caveats from the `sign` method apply to `signPartial`
+   * All the caveats from the `sign` method apply to `partialSign`
    */
-  signPartial(...partialSigners: Array<PublicKey | Account>) {
-    if (partialSigners.length === 0) {
+  partialSign(...signers: Array<Account>) {
+    if (signers.length === 0) {
       throw new Error('No signers');
     }
 
-    function partialSignerPublicKey(accountOrPublicKey: any): PublicKey {
-      if ('publicKey' in accountOrPublicKey) {
-        return accountOrPublicKey.publicKey;
-      }
-      return accountOrPublicKey;
-    }
-
-    function signerAccount(accountOrPublicKey: any): ?Account {
-      if (
-        'publicKey' in accountOrPublicKey &&
-        'secretKey' in accountOrPublicKey
-      ) {
-        return accountOrPublicKey;
-      }
-    }
-
-    const signatures: Array<SignaturePubkeyPair> = partialSigners.map(
-      accountOrPublicKey => ({
-        signature: null,
-        publicKey: partialSignerPublicKey(accountOrPublicKey),
-      }),
-    );
-    this.signatures = signatures;
     const signData = this.serializeMessage();
```

Comment:
@t-nelson see here, we set signature order before serializing the first time... (signPartial was called from sign)